### PR TITLE
PHP 8.1: Enums

### LINKED ISSUE #789: Implement adapters for `ReflectionAttribute`, `ReflectionEnumPureCase`, `ReflectionEnumBackedCase` and `ReflectionEnumUnitcase` when supported by PHP
Ref: https://bugs.php.net/bug.php?id=81474

### LINKED ISSUE #806: PHP 8.1 support
- [x] Green build https://github.com/Roave/BetterReflection/pull/767
- [x] Readonly properties https://github.com/Roave/BetterReflection/pull/811
- [x] Final constants https://github.com/Roave/BetterReflection/pull/814
- [x] `never` type https://github.com/Roave/BetterReflection/pull/812
- [x] `ReflectionClass::isEnum()` and `ReflectionObject::isEnum()` https://github.com/Roave/BetterReflection/pull/813
- [x] `ReflectionFunction::isStatic()` https://github.com/Roave/BetterReflection/pull/815
- [x] ~~`getClosureUsedVariables()`~~
- [x] Intersection types https://github.com/Roave/BetterReflection/pull/818
- [x] Tentative return types https://github.com/Roave/BetterReflection/pull/817
- [x] New in initializers https://github.com/Roave/BetterReflection/pull/819
- [x] Enums https://github.com/Roave/BetterReflection/pull/827

### PR DESCRIPTION
**Pull request by kukulich**
Fixes #789
Fixes #806

### CODE REVIEW

File: src/Reflection/ReflectionEnum.php [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+use Roave\BetterReflection\Reflector\Reflector;
+use Roave\BetterReflection\SourceLocator\Located\LocatedSource;
+
+use function array_combine;
+use function array_filter;
+use function array_key_exists;
+use function array_map;
+use function assert;
+
+class ReflectionEnum extends ReflectionClass
```
Comment (Ocramius) [inline src/Reflection/ReflectionEnum.php:23]: Overall, not a fan of having inheritance here for enums (core has that), but it's not a blocker for 5.x for now.

I'd expect this to probably be broken up in 6.x

File: src/Reflection/ReflectionEnum.php [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+    public function isBacked(): bool
+    {
+        return $this->node->scalarType !== null;
+    }
+
+    public function getBackingType(): ?ReflectionNamedType
+    {
+        if ($this->node->scalarType === null) {
+            return null;
```
Comment (Ocramius) [inline src/Reflection/ReflectionEnum.php]: We should probably not expose a nullable API in our core here, or we should perhaps get rid of `isBacked()` and `hasCase()`.

File: src/Reflection/ReflectionEnum.php [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        }
+
+        $backingType = ReflectionNamedType::createFromTypeAndReflector($this->node->scalarType);
```
Comment (Ocramius) [inline src/Reflection/ReflectionEnum.php]: `createFromTypeAndReflector` is missing a reflector here

**Review by Ocramius — CHANGES_REQUESTED** [review]

File: src/Reflection/ReflectionEnum.php (hunk 3 above)
Comment (kukulich) [reply src/Reflection/ReflectionEnum.php]: https://github.com/Roave/BetterReflection/pull/831

File: src/Reflection/ReflectionEnum.php (hunk 2 above)
Comment (kukulich) [reply src/Reflection/ReflectionEnum.php]: `hasCase`/`getCase` - it's same as `hasProperty`/`getProperty` - I tried to use the same pattern.

`getBackingType` - I modified it to the same pattern as `getDefaultValue()`. It now throws exception when enum is not backed.

**Comment by kukulich** [comment]
@Ocramius @asgrim 

I'm not sure... Should I implement https://3v4l.org/dN6BB as part of `BetterReflectionClass` or only in adapter?

**Comment by Ocramius** [comment]
> or only in adapter?

Looks like an adapter-only concern, since it causes autoloading :-\

**Comment by kukulich** [comment]
@Ocramius I think this return value does not need autoload because `enum(Foo::BOO)` is just string. I'm just not sure if BR enum/class should return enum cases in list of constants.

**Comment by Ocramius** [comment]
```
  ["BOO"]=>
  enum(Foo::BOO)
```

Is that really just a `string`? Looks like the engine is doing more there :thinking:

**Comment by kukulich** [comment]
@Ocramius Oh, shit. You're right https://3v4l.org/fQiIg :(

**Comment by kukulich** [comment]
Btw my question is for `getReflectionConstants()` as well.

**Comment by Ocramius** [comment]
> @Ocramius Oh, shit. You're right https://3v4l.org/fQiIg :(

Yeah, that's kinda what I was referring to - it's actually instantiating the ENUM, so runtime evaluation is quite needed :-\ 

I think it would be potentially acceptable to have that happen in the **adapters**, but not in core (similar to what you did with `new in initializers` in #819 )

**Comment by Ocramius** [comment]
I think besides dependabot, this is the last one we need to push through for `5.0.0` :O

**Comment by kukulich** [comment]
@Ocramius This is already ready for review :) `getConstants()` for enum cases is implemented only in adapters

**Comment by Ocramius** [comment]
@asgrim since I'm off until Wednesday (literally on a trip), can you perhaps help getting this through in the next days? 

Also: let's please not tag `5.0.0` until we tried it with `roave/backward-compatibility-check` (should be few minor changes)

**Comment by asgrim** [comment]
> @asgrim since I'm off until Wednesday (literally on a trip), can you perhaps help getting this through in the next days?
> 
> Also: let's please not tag `5.0.0` until we tried it with `roave/backward-compatibility-check` (should be few minor changes)

I'm also off until Weds :sweat_smile: I'll see if I can get to it but no promises, quite a beast this one @kukulich :grin:

**Comment by kukulich** [comment]
@asgrim There are more tests than the code changes 😃

File: src/Reflection/Adapter/ReflectionClass.php [hunk 4]
```diff
@@ -202,6 +232,16 @@ private function filterBetterReflectionClassConstants(?int $filter): array
...
         }
 
+        if (
+            $this->betterReflectionClass instanceof BetterReflectionEnum
+            && (
+                $filter === null
+                || $filter & ReflectionClassConstant::IS_PUBLIC
+            )
+        ) {
+            $reflectionConstants += $this->betterReflectionClass->getCases();
```
Comment (Ocramius) [inline src/Reflection/Adapter/ReflectionClass.php:242]: Not a fan of `+=`, but we'll make it work :D

**Review by Ocramius — APPROVED** [review]
Absolutely impeccable! :O 

Impressive work, @kukulich, thank you!

**Pull request merged**